WASM support broken in 0.32.1?

I couldn't find the documentation on how to make it work in WASM, and getting this error when trying to build with `wasm-pack`.



```sh
error: The wasm32-unknown-unknown targets are not supported by default; you may need to enable the "wasm_js" configuration flag. Note that enabling the `wasm_js` feature flag alone is insufficient. For more information see: https://docs.rs/getrandom/0.3.3/#webassembly-support
   --> /Users/jeremy/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/getrandom-0.3.3/src/backends.rs:168:9
    |
168 | /         compile_error!(concat!(
169 | |             "The wasm32-unknown-unknown targets are not supported by default; \
170 | |             you may need to enable the \"wasm_js\" configuration flag. Note \
171 | |             that enabling the `wasm_js` feature flag alone is insufficient. \
172 | |             For more information see: \
173 | |             https://docs.rs/getrandom/", env!("CARGO_PKG_VERSION"), "/#webassembly-support"
174 | |         ));
    | |__________^

error[E0425]: cannot find function `fill_inner` in module `backends`
  --> /Users/jeremy/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/getrandom-0.3.3/src/lib.rs:99:19
   |
99 |         backends::fill_inner(dest)?;
   |                   ^^^^^^^^^^ not found in `backends`

error[E0425]: cannot find function `inner_u32` in module `backends`
   --> /Users/jeremy/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/getrandom-0.3.3/src/lib.rs:128:15
    |
128 |     backends::inner_u32()
    |               ^^^^^^^^^ not found in `backends`
    |
help: consider importing this function
    |
33  + use crate::util::inner_u32;
    |
help: if you import `inner_u32`, refer to it directly
    |
128 -     backends::inner_u32()
128 +     inner_u32()
    |

error[E0425]: cannot find function `inner_u64` in module `backends`
   --> /Users/jeremy/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/getrandom-0.3.3/src/lib.rs:142:15
    |
142 |     backends::inner_u64()
    |               ^^^^^^^^^ not found in `backends`
    |
help: consider importing this function
    |
33  + use crate::util::inner_u64;
    |
help: if you import `inner_u64`, refer to it directly
    |
142 -     backends::inner_u64()
142 +     inner_u64()
    |

   Compiling lock_api v0.4.13
   Compiling borrow-or-share v0.2.2
```

Also tried with disabling default features as someone had mentioned, but that leads to:

```sh
jeremy@Jeremys-MacBook-Pro simd-vec-db-web-client % npm run build

> simd-vec-db-web-client@0.1.0-alpha build
> npm run build-rust-wasm && vite build


> simd-vec-db-web-client@0.1.0-alpha build-rust-wasm
> wasm-pack build ./rust/simd-vec-db-web-client --release --target web --out-dir ./pkg

Error: `cargo metadata` exited with an error: error: failed to load manifest for workspace member `/Users/jeremy/Projects/rust-simd-vec-db/simd-vec-db`
referenced by workspace at `/Users/jeremy/Projects/rust-simd-vec-db/Cargo.toml`

Caused by:
  failed to load manifest for dependency `simd-vec-db-document`

Caused by:
  failed to parse manifest at `/Users/jeremy/Projects/rust-simd-vec-db/simd-vec-db-document/Cargo.toml`

Caused by:
  error inheriting `jsonschema` from workspace root manifest's `workspace.dependencies.jsonschema`

Caused by:
  `default-features = false` cannot override workspace's `default-features`

Caused by: `cargo metadata` exited with an error: error: failed to load manifest for workspace member `/Users/jeremy/Projects/rust-simd-vec-db/simd-vec-db`
referenced by workspace at `/Users/jeremy/Projects/rust-simd-vec-db/Cargo.toml`

Caused by:
  failed to load manifest for dependency `simd-vec-db-document`

Caused by:
  failed to parse manifest at `/Users/jeremy/Projects/rust-simd-vec-db/simd-vec-db-document/Cargo.toml`

Caused by:
  error inheriting `jsonschema` from workspace root manifest's `workspace.dependencies.jsonschema`

Caused by:
  `default-features = false` cannot override workspace's `default-features`
```

--- a/.cargo/config.toml
+++ b/.cargo/config.toml
@@ -1,2 +1,5 @@
 [target.wasm32-wasip1]
 runner = "wasmtime"
+
+[target.wasm32-unknown-unknown]
+runner = "wasm-bindgen-test-runner"

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -8,8 +8,8 @@ on:
 
 jobs:
 
-  commitsar:
-    name: Verify commit messages
+  check-commits:
+    name: Check commit messages
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -17,8 +17,8 @@ jobs:
           fetch-depth: 0
       - uses: aevea/commitsar@v0.20.2
 
-  msrv:
-    name: MSRV
+  check-msrv:
+    name: Check MSRV
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -42,7 +42,7 @@ jobs:
       - run: cargo update -p zerofrom --precise 0.1.5
       - run: cargo build -p jsonschema
 
-  test-stable:
+  test-native:
     strategy:
       fail-fast: false
       matrix:
@@ -65,28 +65,88 @@ jobs:
       - run: cargo test --no-fail-fast --all-features
 
   test-wasm:
-    name: Test on WASM
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - target: wasm32-wasip1
+            needs_wasmtime: true
+            needs_wasm_bindgen: false
+          - target: wasm32-unknown-unknown
+            needs_wasmtime: false
+            needs_wasm_bindgen: true
+    name: Test on ${{ matrix.target }}
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
         with:
           submodules: true
 
-      - name: Setup `wasmtime`
-        uses: bytecodealliance/actions/wasmtime/setup@v1
-
       - uses: dtolnay/rust-toolchain@stable
         with:
-          targets: 'wasm32-wasip1'
+          targets: ${{ matrix.target }}
 
       - uses: Swatinem/rust-cache@v2
         with:
           cache-all-crates: "true"
 
-      - run: cargo test --target wasm32-wasip1 --no-default-features -p jsonschema
+      - name: Setup `wasmtime`
+        if: ${{ matrix.needs_wasmtime }}
+        uses: bytecodealliance/actions/wasmtime/setup@v1
 
-  coverage:
-    name: Run test coverage
+      - name: Install wasm-bindgen
+        if: ${{ matrix.needs_wasm_bindgen }}
+        uses: taiki-e/install-action@wasm-bindgen
+        with:
+          tool: 'wasm-bindgen@0.2.105'
+
+      - name: Run tests
+        run: cargo test --target ${{ matrix.target }} --no-default-features -p jsonschema
+
+  check-wasm-compile-errors:
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - target: wasm32-unknown-unknown
+            feature: resolve-http
+            pattern: "resolve-http.*resolve-file.*not supported.*wasm32-unknown-unknown"
+          - target: wasm32-unknown-unknown
+            feature: resolve-file
+            pattern: "resolve-http.*resolve-file.*not supported.*wasm32-unknown-unknown"
+          - target: wasm32-unknown-unknown
+            feature: resolve-async
+            pattern: "resolve-async.*not supported.*wasm32"
+          - target: wasm32-wasip1
+            feature: resolve-async
+            pattern: "resolve-async.*not supported.*wasm32"
+    name: Check ${{ matrix.feature }} fails on ${{ matrix.target }}
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v5
+
+      - uses: dtolnay/rust-toolchain@stable
+        with:
+          targets: ${{ matrix.target }}
+
+      - uses: Swatinem/rust-cache@v2
+        with:
+          cache-all-crates: "true"
+          key: ${{ matrix.target }}-${{ matrix.feature }}
+
+      - name: Verify compile error
+        run: |
+          output=$(cargo check --target ${{ matrix.target }} --features ${{ matrix.feature }} -p jsonschema 2>&1 || true)
+          echo "$output"
+          if ! echo "$output" | grep -q "${{ matrix.pattern }}"; then
+            echo "Expected compile error not found!"
+            echo "Expected pattern: ${{ matrix.pattern }}"
+            exit 1
+          fi
+          echo "✓ Correctly rejects ${{ matrix.feature }} on ${{ matrix.target }}"
+
+  test-coverage:
+    name: Test coverage
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -116,7 +176,7 @@ jobs:
           files: lcov.info
           fail_ci_if_error: true
 
-  miri:
+  test-miri:
     name: Test with Miri
     runs-on: ubuntu-22.04
     steps:
@@ -135,8 +195,8 @@ jobs:
       - name: Run tests with Miri
         run: cargo miri test -p referencing --lib
 
-  lints-python:
-    name: Python lints
+  lint-python:
+    name: Lint Python
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -156,7 +216,7 @@ jobs:
         os: [ubuntu-22.04, macos-15, windows-2022]
         python-version: ['3.8', '3.9', '3.10', '3.11', '3.12', '3.13']
 
-    name: Python ${{ matrix.python-version }} on ${{ matrix.os }}
+    name: Test Python ${{ matrix.python-version }} on ${{ matrix.os }}
     runs-on: ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v5
@@ -172,8 +232,8 @@ jobs:
       - run: uvx --with=".[tests]" --python='${{ matrix.python-version }}' pytest tests-py
         working-directory: crates/jsonschema-py
 
-  fmt:
-    name: Rustfmt
+  lint-fmt:
+    name: Lint formatting
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -184,8 +244,8 @@ jobs:
 
       - run: cargo fmt --all -- --check
 
-  lints-rust:
-    name: Rust lints
+  lint-clippy:
+    name: Lint Clippy
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -197,6 +257,8 @@ jobs:
           components: clippy
 
       - uses: Swatinem/rust-cache@v2
+        with:
+          cache-all-crates: "true"
 
       - run: cargo clippy --all-targets --all-features -- -D warnings
       - run: cargo clippy --all-targets --all-features -- -D warnings
@@ -204,8 +266,12 @@ jobs:
       - run: cargo clippy --all-targets --all-features -- -D warnings
         working-directory: fuzz
 
-  lints-wasm:
-    name: Rust lints on WASM
+  lint-clippy-wasm:
+    strategy:
+      fail-fast: false
+      matrix:
+        target: [wasm32-unknown-unknown, wasm32-wasip1]
+    name: Lint Clippy on ${{ matrix.target }}
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -215,16 +281,18 @@ jobs:
       - uses: dtolnay/rust-toolchain@stable
         with:
           components: clippy
-          targets: 'wasm32-wasip1'
+          targets: ${{ matrix.target }}
 
       - uses: Swatinem/rust-cache@v2
         with:
           cache-all-crates: "true"
+          key: ${{ matrix.target }}
 
-      - run: cargo clippy --target wasm32-wasip1 --no-default-features --all-targets -p jsonschema
+      - run: cargo clippy --target ${{ matrix.target }} --all-targets --no-default-features -p jsonschema -- -D warnings
+      - run: cargo clippy --target ${{ matrix.target }} --all-targets --no-default-features -p referencing -- -D warnings
 
-  check-releases-rust:
-    name: Check Rust releases
+  check-release-rust:
+    name: Check Rust release
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -234,13 +302,15 @@ jobs:
       - uses: dtolnay/rust-toolchain@stable
 
       - uses: Swatinem/rust-cache@v2
+        with:
+          cache-all-crates: "true"
 
       - run: cargo publish --dry-run -p referencing
       - run: cargo publish --dry-run -p jsonschema
       - run: cargo publish --dry-run -p jsonschema-cli
 
-  check-releases-python:
-    name: Check Python releases
+  check-release-python:
+    name: Check Python release
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -271,8 +341,21 @@ jobs:
           uv pip install dist/*.whl --force-reinstall
         working-directory: crates/jsonschema-py
 
-  features:
-    name: Check features
+  check-features:
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - target: x86_64-unknown-linux-gnu
+            name: native
+            flags: "--feature-powerset"
+          - target: wasm32-unknown-unknown
+            name: wasm32-unknown-unknown
+            flags: "--feature-powerset --exclude-features default,resolve-http,resolve-file,resolve-async,retrieve-async"
+          - target: wasm32-wasip1
+            name: wasm32-wasip1
+            flags: "--feature-powerset --exclude-features resolve-async,retrieve-async"
+    name: Check features on ${{ matrix.name }}
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -280,19 +363,27 @@ jobs:
           submodules: true
 
       - uses: dtolnay/rust-toolchain@stable
+        with:
+          targets: ${{ matrix.target }}
 
       - uses: Swatinem/rust-cache@v2
+        with:
+          key: ${{ matrix.target }}
 
       - uses: taiki-e/install-action@cargo-hack
 
-      - run: cargo hack check --feature-powerset
+      - name: Check features for jsonschema
+        run: cargo hack check --target ${{ matrix.target }} ${{ matrix.flags }} -p jsonschema
+
+      - name: Check features for referencing
+        run: cargo hack check --target ${{ matrix.target }} ${{ matrix.flags }} -p referencing
 
   fuzz:
-    name: Fuzzing ${{ matrix.target }}
     strategy:
       fail-fast: false
       matrix:
         target: ["builder", "validation", "referencing"]
+    name: Fuzz ${{ matrix.target }}
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5
@@ -307,8 +398,8 @@ jobs:
 
       - run: cargo +nightly fuzz run --release ${{ matrix.target }} fuzz/seeds -- -dict=fuzz/dict -max_total_time=60
 
-  spell-check:
-    name: Spell Check
+  lint-spelling:
+    name: Lint spelling
     runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v5

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,7 @@
 ### Fixed
 
 - Hostname and IDN hostname formats now decode `xn--` labels, reject leading combining marks/uppercase prefixes, and enforce the latest JSON Schema punycode context rules.
+- Restore `wasm32-unknown-unknown` support. [#785](https://github.com/Stranger6667/jsonschema/issues/785)
 
 ### Performance
 

--- a/Justfile
+++ b/Justfile
@@ -34,3 +34,15 @@ bench-py *FLAGS:
 
 miri:
   cargo +nightly miri test -p referencing
+
+install-wasm-deps:
+  rustup target add wasm32-wasip1 wasm32-unknown-unknown
+  cargo install wasm-bindgen-cli --vers 0.2.105
+
+test-wasm32-wasip1:
+  cargo test --target wasm32-wasip1 --no-default-features -p jsonschema
+
+test-wasm32-unknown-unknown:
+  cargo test --target wasm32-unknown-unknown --no-default-features -p jsonschema
+
+test-wasm: test-wasm32-wasip1 test-wasm32-unknown-unknown

--- a/README.md
+++ b/README.md
@@ -86,6 +86,36 @@ You can check the current status on the [Bowtie Report](https://bowtie.report/#/
 
 For detailed benchmarks, see our [full performance comparison](https://github.com/Stranger6667/jsonschema/tree/master/crates/benchmark-suite).
 
+## WebAssembly
+
+`jsonschema` supports WebAssembly with different capabilities depending on the target:
+
+### Browser/JavaScript (`wasm32-unknown-unknown`)
+
+For browser environments, use without default features:
+
+```toml
+jsonschema = { version = "0.33", default-features = false }
+```
+
+External references (`resolve-http`, `resolve-file`) are not supported. Implement a custom retriever for external schemas (see [documentation](https://docs.rs/jsonschema) for details).
+
+### WASI (`wasm32-wasip1`)
+
+WASI environments have limited built-in support due to library architecture:
+
+**Supported:**
+- Synchronous file resolution
+- Custom synchronous retrievers
+
+```toml
+jsonschema = { version = "0.33", default-features = false, features = ["resolve-file"] }
+```
+
+**Not supported:** Built-in HTTP and async resolution (requires `Send + Sync` trait bounds incompatible with single-threaded WASM).
+
+**Workaround:** Implement a custom synchronous retriever that wraps your async HTTP client internally.
+
 ## Minimum Supported Rust Version (MSRV)
 
 This crate requires Rust 1.83.0 or later.

--- a/crates/jsonschema-referencing/Cargo.toml
+++ b/crates/jsonschema-referencing/Cargo.toml
@@ -13,6 +13,9 @@ license.workspace = true
 default = []
 retrieve-async = ["dep:async-trait", "dep:futures"]
 
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+getrandom = { version = "0.3.4", features = ["wasm_js"] }
+
 [dependencies]
 ahash.workspace = true
 
@@ -27,6 +30,8 @@ serde_json.workspace = true
 benchmark = { path = "../benchmark/" }
 codspeed-criterion-compat = { version = "4.1", default-features = false }
 criterion = { version = "0.7", default-features = false }
+# This one is to make `wasm32-unknown-unknown` lints happy (codspeed depends on it)
+getrandom = { version = "0.2", features = ["js"] }
 referencing_testsuite = { package = "jsonschema-referencing-testsuite", path = "../jsonschema-referencing-testsuite/" }
 test-case = "3.3.1"
 tokio = { version = "1", features = ["macros", "rt"] }

--- a/crates/jsonschema/Cargo.toml
+++ b/crates/jsonschema/Cargo.toml
@@ -13,7 +13,6 @@ authors.workspace = true
 
 [features]
 default = ["resolve-http", "resolve-file"]
-
 resolve-http = ["reqwest"]
 resolve-file = []
 resolve-async = [
@@ -22,6 +21,18 @@ resolve-async = [
     "dep:async-trait",
     "dep:tokio",
 ]
+
+[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
+reqwest = { version = "0.12", features = [
+  "blocking",
+  "json",
+], default-features = false, optional = true }
+tokio = { version = "1.0", features = ["fs", "rt"], optional = true }
+
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+getrandom = { version = "0.3.4", features = ["wasm_js"] }
+reqwest = { version = "0.12", features = ["json"], default-features = false, optional = true }
+tokio = { version = "1.0", features = ["rt"], optional = true }
 
 [dependencies]
 ahash.workspace = true
@@ -41,28 +52,26 @@ percent-encoding = "2.3"
 referencing = { version = "0.33.0", path = "../jsonschema-referencing" }
 regex = "1.11.1"
 regex-syntax = "0.8.5"
-reqwest = { version = "0.12", features = [
-  "blocking",
-  "json",
-], default-features = false, optional = true }
 serde.workspace = true
 serde_json.workspace = true
-
-tokio = { version = "1.0", features = ["fs", "rt"], optional = true }
 unicode-general-category = "1.1"
-uuid-simd = "0.8"
+uuid-simd = { version = "0.8", default-features = false, features = ["std", "detect"] }
 
 [dev-dependencies]
+test-case = "3"
+
+[target.'cfg(all(target_arch = "wasm32", target_os = "unknown"))'.dev-dependencies]
+wasm-bindgen-test = "0.3.55"
+
+[target.'cfg(not(target_arch = "wasm32"))'.dev-dependencies]
 benchmark = { path = "../benchmark/" }
 codspeed-criterion-compat = { version = "4.1", default-features = false }
 criterion = { version = "0.7", default-features = false }
+mockito = "1.5"
+tempfile = "3.13.0"
 test-case = "3"
 testsuite = { package = "jsonschema-testsuite", path = "../jsonschema-testsuite" }
 tokio = { version = "1", features = ["macros", "rt"] }
-
-[target.'cfg(not(target_arch = "wasm32"))'.dev-dependencies]
-mockito = "1.5"
-tempfile = "3.13.0"
 
 [lints.clippy]
 result_large_err = "allow"

--- a/crates/jsonschema/benches/errors.rs
+++ b/crates/jsonschema/benches/errors.rs
@@ -1,23 +1,31 @@
-use benchmark::run_error_formatting_benchmarks;
-use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
-use serde_json::Value;
+#[cfg(not(target_arch = "wasm32"))]
+mod bench {
+    pub use benchmark::run_error_formatting_benchmarks;
+    pub use criterion::{criterion_group, BenchmarkId, Criterion};
+    pub use serde_json::Value;
 
-fn bench_error_formatting(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
-    let validator = jsonschema::validator_for(schema).expect("Valid schema");
-    let error = validator.validate(instance).unwrap_err();
+    pub fn bench_error_formatting(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
+        let validator = jsonschema::validator_for(schema).expect("Valid schema");
+        let error = validator.validate(instance).unwrap_err();
 
-    c.bench_with_input(
-        BenchmarkId::new("error_formatting", name),
-        &error,
-        |b, error| b.iter_with_large_drop(|| error.to_string()),
-    );
+        c.bench_with_input(
+            BenchmarkId::new("error_formatting", name),
+            &error,
+            |b, error| b.iter_with_large_drop(|| error.to_string()),
+        );
+    }
+
+    pub fn run_benchmarks(c: &mut Criterion) {
+        run_error_formatting_benchmarks(&mut |name, schema, instance| {
+            bench_error_formatting(c, name, schema, instance);
+        });
+    }
+
+    criterion_group!(error_formatting, run_benchmarks);
 }
 
-fn run_benchmarks(c: &mut Criterion) {
-    run_error_formatting_benchmarks(&mut |name, schema, instance| {
-        bench_error_formatting(c, name, schema, instance);
-    });
-}
+#[cfg(not(target_arch = "wasm32"))]
+criterion::criterion_main!(bench::error_formatting);
 
-criterion_group!(error_formatting, run_benchmarks);
-criterion_main!(error_formatting);
+#[cfg(target_arch = "wasm32")]
+fn main() {}

--- a/crates/jsonschema/benches/jsonschema.rs
+++ b/crates/jsonschema/benches/jsonschema.rs
@@ -1,59 +1,67 @@
-use benchmark::Benchmark;
-use codspeed_criterion_compat::{criterion_group, criterion_main, BenchmarkId, Criterion};
-use serde_json::Value;
+#[cfg(not(target_arch = "wasm32"))]
+mod bench {
+    pub use benchmark::Benchmark;
+    pub use codspeed_criterion_compat::{criterion_group, BenchmarkId, Criterion};
+    pub use serde_json::Value;
 
-fn bench_build(c: &mut Criterion, name: &str, schema: &Value) {
-    c.bench_with_input(BenchmarkId::new("build", name), schema, |b, schema| {
-        b.iter_with_large_drop(|| jsonschema::validator_for(schema).expect("Valid schema"));
-    });
-}
-
-fn bench_is_valid(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
-    let validator = jsonschema::validator_for(schema).expect("Valid schema");
-    c.bench_with_input(
-        BenchmarkId::new("is_valid", name),
-        instance,
-        |b, instance| {
-            b.iter(|| {
-                let _ = validator.is_valid(instance);
-            });
-        },
-    );
-}
-
-fn bench_validate(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
-    let validator = jsonschema::validator_for(schema).expect("Valid schema");
-    c.bench_with_input(
-        BenchmarkId::new("validate", name),
-        instance,
-        |b, instance| {
-            b.iter(|| {
-                let _ = validator.validate(instance);
-            });
-        },
-    );
-}
-
-fn bench_apply(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
-    let validator = jsonschema::validator_for(schema).expect("Valid schema");
-    c.bench_with_input(BenchmarkId::new("apply", name), instance, |b, instance| {
-        b.iter_with_large_drop(|| validator.apply(instance).basic());
-    });
-}
-
-fn run_benchmarks(c: &mut Criterion) {
-    for benchmark in Benchmark::iter() {
-        benchmark.run(&mut |name, schema, instances| {
-            bench_build(c, name, schema);
-            for instance in instances {
-                let name = format!("{}/{}", name, instance.name);
-                bench_is_valid(c, &name, schema, &instance.data);
-                bench_validate(c, &name, schema, &instance.data);
-                bench_apply(c, &name, schema, &instance.data);
-            }
+    pub fn bench_build(c: &mut Criterion, name: &str, schema: &Value) {
+        c.bench_with_input(BenchmarkId::new("build", name), schema, |b, schema| {
+            b.iter_with_large_drop(|| jsonschema::validator_for(schema).expect("Valid schema"));
         });
     }
+
+    pub fn bench_is_valid(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
+        let validator = jsonschema::validator_for(schema).expect("Valid schema");
+        c.bench_with_input(
+            BenchmarkId::new("is_valid", name),
+            instance,
+            |b, instance| {
+                b.iter(|| {
+                    let _ = validator.is_valid(instance);
+                });
+            },
+        );
+    }
+
+    pub fn bench_validate(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
+        let validator = jsonschema::validator_for(schema).expect("Valid schema");
+        c.bench_with_input(
+            BenchmarkId::new("validate", name),
+            instance,
+            |b, instance| {
+                b.iter(|| {
+                    let _ = validator.validate(instance);
+                });
+            },
+        );
+    }
+
+    pub fn bench_apply(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
+        let validator = jsonschema::validator_for(schema).expect("Valid schema");
+        c.bench_with_input(BenchmarkId::new("apply", name), instance, |b, instance| {
+            b.iter_with_large_drop(|| validator.apply(instance).basic());
+        });
+    }
+
+    pub fn run_benchmarks(c: &mut Criterion) {
+        for benchmark in Benchmark::iter() {
+            benchmark.run(&mut |name, schema, instances| {
+                bench_build(c, name, schema);
+                for instance in instances {
+                    let name = format!("{}/{}", name, instance.name);
+                    bench_is_valid(c, &name, schema, &instance.data);
+                    bench_validate(c, &name, schema, &instance.data);
+                    bench_apply(c, &name, schema, &instance.data);
+                }
+            });
+        }
+    }
+
+    criterion_group!(jsonschema, run_benchmarks);
 }
 
-criterion_group!(jsonschema, run_benchmarks);
-criterion_main!(jsonschema);
+#[cfg(not(target_arch = "wasm32"))]
+codspeed_criterion_compat::criterion_main!(bench::jsonschema);
+
+#[cfg(target_arch = "wasm32")]
+fn main() {}

--- a/crates/jsonschema/benches/keywords.rs
+++ b/crates/jsonschema/benches/keywords.rs
@@ -1,56 +1,64 @@
-use benchmark::run_keyword_benchmarks;
-use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
-use serde_json::Value;
+#[cfg(not(target_arch = "wasm32"))]
+mod bench {
+    pub use benchmark::run_keyword_benchmarks;
+    pub use criterion::{criterion_group, BenchmarkId, Criterion};
+    pub use serde_json::Value;
 
-fn validator_for(schema: &Value) -> jsonschema::Validator {
-    jsonschema::options()
-        .with_draft(jsonschema::Draft::Draft7)
-        .build(schema)
-        .expect("Schema used in benchmarks should compile")
+    pub fn validator_for(schema: &Value) -> jsonschema::Validator {
+        jsonschema::options()
+            .with_draft(jsonschema::Draft::Draft7)
+            .build(schema)
+            .expect("Schema used in benchmarks should compile")
+    }
+
+    pub fn bench_keyword_build(c: &mut Criterion, name: &str, schema: &Value) {
+        c.bench_function(&format!("keyword/{name}/build"), |b| {
+            b.iter_with_large_drop(|| validator_for(schema));
+        });
+    }
+
+    pub fn bench_keyword_is_valid(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
+        let validator = validator_for(schema);
+        c.bench_with_input(
+            BenchmarkId::new(format!("keyword/{name}"), "is_valid"),
+            instance,
+            |b, instance| {
+                b.iter(|| {
+                    let _ = validator.is_valid(instance);
+                });
+            },
+        );
+    }
+
+    pub fn bench_keyword_validate(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
+        let validator = validator_for(schema);
+        c.bench_with_input(
+            BenchmarkId::new(format!("keyword/{name}"), "validate"),
+            instance,
+            |b, instance| {
+                b.iter(|| {
+                    let _ = validator.validate(instance);
+                });
+            },
+        );
+    }
+
+    pub fn run_benchmarks(c: &mut Criterion) {
+        run_keyword_benchmarks(&mut |name, schema, instances| {
+            bench_keyword_build(c, name, schema);
+            for instance in instances {
+                let name = format!("jsonschema/{}/{}", name, instance.name);
+                bench_keyword_is_valid(c, &name, schema, &instance.data);
+                bench_keyword_validate(c, &name, schema, &instance.data);
+            }
+        });
+    }
+
+    criterion_group!(keywords, run_benchmarks);
 }
 
-fn bench_keyword_build(c: &mut Criterion, name: &str, schema: &Value) {
-    c.bench_function(&format!("keyword/{name}/build"), |b| {
-        b.iter_with_large_drop(|| validator_for(schema));
-    });
-}
+#[cfg(not(target_arch = "wasm32"))]
+criterion::criterion_main!(bench::keywords);
 
-fn bench_keyword_is_valid(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
-    let validator = validator_for(schema);
-    c.bench_with_input(
-        BenchmarkId::new(format!("keyword/{name}"), "is_valid"),
-        instance,
-        |b, instance| {
-            b.iter(|| {
-                let _ = validator.is_valid(instance);
-            });
-        },
-    );
-}
-
-fn bench_keyword_validate(c: &mut Criterion, name: &str, schema: &Value, instance: &Value) {
-    let validator = validator_for(schema);
-    c.bench_with_input(
-        BenchmarkId::new(format!("keyword/{name}"), "validate"),
-        instance,
-        |b, instance| {
-            b.iter(|| {
-                let _ = validator.validate(instance);
-            });
-        },
-    );
-}
-
-fn run_benchmarks(c: &mut Criterion) {
-    run_keyword_benchmarks(&mut |name, schema, instances| {
-        bench_keyword_build(c, name, schema);
-        for instance in instances {
-            let name = format!("jsonschema/{}/{}", name, instance.name);
-            bench_keyword_is_valid(c, &name, schema, &instance.data);
-            bench_keyword_validate(c, &name, schema, &instance.data);
-        }
-    });
-}
-
-criterion_group!(keywords, run_benchmarks);
-criterion_main!(keywords);
+#[cfg(target_arch = "wasm32")]
+fn main() {}

--- a/crates/jsonschema/benches/location.rs
+++ b/crates/jsonschema/benches/location.rs
@@ -1,54 +1,61 @@
-use std::hint::black_box;
+#[cfg(not(target_arch = "wasm32"))]
+mod bench {
+    pub use criterion::{criterion_group, BenchmarkId, Criterion};
+    pub use jsonschema::paths::{Location, LocationSegment};
+    pub use std::hint::black_box;
 
-use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
-use jsonschema::paths::{Location, LocationSegment};
+    pub fn benchmark_into_iterator(c: &mut Criterion) {
+        let empty = Location::new();
+        let small = vec!["a", "b", "c"]
+            .into_iter()
+            .map(LocationSegment::from)
+            .chain((0..3).map(LocationSegment::from))
+            .collect::<Location>();
+        let large = (0..500)
+            .map(|_| LocationSegment::from("abc"))
+            .chain((0..500).map(LocationSegment::from))
+            .collect::<Location>();
 
-fn benchmark_into_iterator(c: &mut Criterion) {
-    let empty = Location::new();
-    let small = vec!["a", "b", "c"]
-        .into_iter()
-        .map(LocationSegment::from)
-        .chain((0..3).map(LocationSegment::from))
-        .collect::<Location>();
-    let large = (0..500)
-        .map(|_| LocationSegment::from("abc"))
-        .chain((0..500).map(LocationSegment::from))
-        .collect::<Location>();
-
-    for (parameter, input) in [("empty", empty), ("small", small), ("large", large)] {
-        c.bench_with_input(
-            BenchmarkId::new("IntoIterator", parameter),
-            &input,
-            |b, i| b.iter_with_large_drop(|| black_box(i).into_iter().collect::<Vec<_>>()),
-        );
+        for (parameter, input) in [("empty", empty), ("small", small), ("large", large)] {
+            c.bench_with_input(
+                BenchmarkId::new("IntoIterator", parameter),
+                &input,
+                |b, i| b.iter_with_large_drop(|| black_box(i).into_iter().collect::<Vec<_>>()),
+            );
+        }
     }
+
+    pub fn benchmark_from_iterator(c: &mut Criterion) {
+        let empty = vec![];
+        let small = vec![
+            LocationSegment::from("a"),
+            LocationSegment::from("b"),
+            LocationSegment::from("c"),
+        ];
+        let large = (0..1000)
+            .map(|_| LocationSegment::from("abc"))
+            .collect::<Vec<_>>();
+
+        for (parameter, input) in [("empty", empty), ("small", small), ("large", large)] {
+            c.bench_with_input(
+                BenchmarkId::new("FromIterator", parameter),
+                &input,
+                |b, i| {
+                    b.iter_batched(
+                        || i.clone().into_iter(),
+                        |i| black_box(i).collect::<Location>(),
+                        criterion::BatchSize::SmallInput,
+                    );
+                },
+            );
+        }
+    }
+
+    criterion_group!(benches, benchmark_into_iterator, benchmark_from_iterator);
 }
 
-fn benchmark_from_iterator(c: &mut Criterion) {
-    let empty = vec![];
-    let small = vec![
-        LocationSegment::from("a"),
-        LocationSegment::from("b"),
-        LocationSegment::from("c"),
-    ];
-    let large = (0..1000)
-        .map(|_| LocationSegment::from("abc"))
-        .collect::<Vec<_>>();
+#[cfg(not(target_arch = "wasm32"))]
+criterion::criterion_main!(bench::benches);
 
-    for (parameter, input) in [("empty", empty), ("small", small), ("large", large)] {
-        c.bench_with_input(
-            BenchmarkId::new("FromIterator", parameter),
-            &input,
-            |b, i| {
-                b.iter_batched(
-                    || i.clone().into_iter(),
-                    |i| black_box(i).collect::<Location>(),
-                    criterion::BatchSize::SmallInput,
-                );
-            },
-        );
-    }
-}
-
-criterion_group!(benches, benchmark_into_iterator, benchmark_from_iterator);
-criterion_main!(benches);
+#[cfg(target_arch = "wasm32")]
+fn main() {}

--- a/crates/jsonschema/src/lib.rs
+++ b/crates/jsonschema/src/lib.rs
@@ -628,19 +628,68 @@
 //!
 //! # WebAssembly support
 //!
-//! When using `jsonschema` in WASM environments, be aware that external references are
-//! not supported by default due to WASM limitations:
-//!    - No filesystem access (`resolve-file` feature)
-//!    - No direct HTTP requests, at least right now (`resolve-http` feature)
+//! `jsonschema` supports WebAssembly with different capabilities based on the target platform:
 //!
-//! To use `jsonschema` in WASM, disable default features:
+//! ## Browser/JavaScript (`wasm32-unknown-unknown`)
+//!
+//! When targeting browser or JavaScript environments, external reference resolution is not
+//! supported by default due to platform limitations:
+//!    - No filesystem access (`resolve-file` feature is not available)
+//!    - No synchronous HTTP requests (`resolve-http` feature is not available)
+//!
+//! To use `jsonschema` in these environments, disable default features:
 //!
 //! ```toml
 //! jsonschema = { version = "x.y.z", default-features = false }
 //! ```
 //!
-//! For external references in WASM you may want to implement a custom retriever.
-//! See the [External References](#external-references) section for implementation details.
+//! Note: Attempting to compile with `resolve-http` or `resolve-file` features on
+//! `wasm32-unknown-unknown` will result in a compile error.
+//!
+//! For external references in browser environments, implement a custom retriever that uses
+//! browser APIs (like `fetch`). See the [External References](#external-references) section.
+//!
+//! ## WASI (`wasm32-wasip1`)
+//!
+//! WASI environments have limited support due to this library's architecture:
+//!
+//! **Supported:**
+//! - Synchronous file resolution (`resolve-file` feature)
+//! - Custom synchronous retrievers (including wrapping async operations)
+//!
+//! **Not Supported:**
+//! - Built-in HTTP resolution - reqwest doesn't support blocking I/O on WASM
+//! - Built-in async resolution - our `AsyncRetrieve` trait requires `Send + Sync` bounds for
+//!   thread-safety in multi-threaded environments, which are incompatible with single-threaded
+//!   WASM futures
+//!
+//! ```toml
+//! jsonschema = { version = "x.y.z", default-features = false, features = ["resolve-file"] }
+//! ```
+//!
+//! **Workaround for HTTP:** Implement a custom synchronous [`Retrieve`] that internally uses
+//! your preferred async HTTP client. Since WASI is single-threaded, you can block on futures
+//! within your retriever implementation.
+
+#[cfg(all(
+    target_arch = "wasm32",
+    target_os = "unknown",
+    any(feature = "resolve-file", feature = "resolve-http")
+))]
+compile_error!(
+    "Features 'resolve-http' and 'resolve-file' are not supported on wasm32-unknown-unknown"
+);
+
+#[cfg(all(target_arch = "wasm32", feature = "resolve-async"))]
+compile_error!(
+    "Feature 'resolve-async' is not supported on wasm32 targets.\n\
+    \n\
+    The AsyncRetrieve trait requires Send + Sync bounds for thread-safety in multi-threaded \n\
+    environments, which are incompatible with single-threaded WASM futures.\n\
+    \n\
+    Workaround: Implement a custom synchronous Retrieve trait that wraps async operations internally.\n\
+    See the documentation for examples: https://docs.rs/jsonschema"
+);
 
 pub(crate) mod compiler;
 mod content_encoding;
@@ -678,12 +727,6 @@ pub use validator::Validator;
 pub use referencing::AsyncRetrieve;
 
 use serde_json::Value;
-
-#[cfg(all(
-    target_arch = "wasm32",
-    any(feature = "resolve-http", feature = "resolve-file")
-))]
-compile_error!("Features 'resolve-http' and 'resolve-file' are not supported on WASM.");
 
 /// Validate `instance` against `schema` and get a `true` if the instance is valid and `false`
 /// otherwise. Draft is detected automatically.

--- a/crates/jsonschema/src/retriever.rs
+++ b/crates/jsonschema/src/retriever.rs
@@ -10,22 +10,28 @@ impl Retrieve for DefaultRetriever {
         &self,
         uri: &Uri<String>,
     ) -> Result<Value, Box<dyn std::error::Error + Send + Sync>> {
-        #[cfg(target_arch = "wasm32")]
+        #[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
         {
-            Err("External references are not supported in WASM".into())
+            Err("External references are not supported on wasm32-unknown-unknown".into())
         }
-        #[cfg(not(target_arch = "wasm32"))]
+        #[cfg(not(all(target_arch = "wasm32", target_os = "unknown")))]
         match uri.scheme().as_str() {
             "http" | "https" => {
-                #[cfg(any(feature = "resolve-http", test))]
+                #[cfg(all(feature = "resolve-http", not(target_arch = "wasm32")))]
                 {
                     Ok(reqwest::blocking::get(uri.as_str())?.json()?)
                 }
-                #[cfg(not(any(feature = "resolve-http", test)))]
-                Err("`resolve-http` feature or a custom resolver is required to resolve external schemas via HTTP".into())
+                #[cfg(all(feature = "resolve-http", target_arch = "wasm32"))]
+                {
+                    Err("Synchronous HTTP retrieval is not supported on wasm32 targets. Use async_validator_for with the resolve-async feature instead".into())
+                }
+                #[cfg(not(feature = "resolve-http"))]
+                {
+                    Err("`resolve-http` feature or a custom resolver is required to resolve external schemas via HTTP".into())
+                }
             }
             "file" => {
-                #[cfg(any(feature = "resolve-file", test))]
+                #[cfg(feature = "resolve-file")]
                 {
                     let path = uri.path().as_str();
                     let path = {
@@ -43,7 +49,7 @@ impl Retrieve for DefaultRetriever {
                     let file = std::fs::File::open(path)?;
                     Ok(serde_json::from_reader(file)?)
                 }
-                #[cfg(not(any(feature = "resolve-file", test)))]
+                #[cfg(not(feature = "resolve-file"))]
                 {
                     Err("`resolve-file` feature or a custom resolver is required to resolve external schemas via files".into())
                 }
@@ -60,22 +66,22 @@ impl referencing::AsyncRetrieve for DefaultRetriever {
         &self,
         uri: &Uri<String>,
     ) -> Result<Value, Box<dyn std::error::Error + Send + Sync>> {
-        #[cfg(target_arch = "wasm32")]
+        #[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
         {
-            Err("External references are not supported in WASM".into())
+            Err("External references are not supported on wasm32-unknown-unknown".into())
         }
-        #[cfg(not(target_arch = "wasm32"))]
+        #[cfg(not(all(target_arch = "wasm32", target_os = "unknown")))]
         match uri.scheme().as_str() {
             "http" | "https" => {
-                #[cfg(any(feature = "resolve-http", test))]
+                #[cfg(feature = "resolve-http")]
                 {
                     Ok(reqwest::get(uri.as_str()).await?.json().await?)
                 }
-                #[cfg(not(any(feature = "resolve-http", test)))]
+                #[cfg(not(feature = "resolve-http"))]
                 Err("`resolve-http` feature or a custom resolver is required to resolve external schemas via HTTP".into())
             }
             "file" => {
-                #[cfg(any(feature = "resolve-file", test))]
+                #[cfg(feature = "resolve-file")]
                 {
                     // File operations are blocking, so we use tokio's spawn_blocking
                     let path = uri.path().as_str().to_string();
@@ -99,7 +105,7 @@ impl referencing::AsyncRetrieve for DefaultRetriever {
                     .await??;
                     Ok(contents)
                 }
-                #[cfg(not(any(feature = "resolve-file", test)))]
+                #[cfg(not(feature = "resolve-file"))]
                 {
                     Err("`resolve-file` feature or a custom resolver is required to resolve external schemas via files".into())
                 }
@@ -187,7 +193,7 @@ mod tests {
     use std::io::Write;
 
     #[test]
-    #[cfg(not(target_arch = "wasm32"))]
+    #[cfg(all(not(target_arch = "wasm32"), feature = "resolve-file"))]
     fn test_retrieve_from_file() {
         let mut temp_file = tempfile::NamedTempFile::new().expect("Failed to create temp file");
         let external_schema = json!({
@@ -230,10 +236,10 @@ mod tests {
 
         assert!(result.is_err());
         let error = result.unwrap_err().to_string();
-        #[cfg(not(target_arch = "wasm32"))]
+        #[cfg(not(all(target_arch = "wasm32", target_os = "unknown")))]
         assert!(error.contains("Unknown scheme"));
-        #[cfg(target_arch = "wasm32")]
-        assert!(error.contains("External references are not supported in WASM"));
+        #[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
+        assert!(error.contains("External references are not supported on wasm32-unknown-unknown"));
     }
 
     #[cfg(not(target_arch = "wasm32"))]
@@ -244,7 +250,7 @@ mod tests {
     }
 
     #[test]
-    #[cfg(not(target_arch = "wasm32"))]
+    #[cfg(all(not(target_arch = "wasm32"), feature = "resolve-file"))]
     fn test_with_base_uri_resolution() {
         let dir = tempfile::tempdir().unwrap();
 
@@ -294,6 +300,7 @@ mod async_tests {
     use std::io::Write;
 
     #[tokio::test]
+    #[cfg(feature = "resolve-file")]
     async fn test_async_retrieve_from_file() {
         let mut temp_file = tempfile::NamedTempFile::new().expect("Failed to create temp file");
         let external_schema = json!({
@@ -359,6 +366,7 @@ mod async_tests {
     }
 
     #[tokio::test]
+    #[cfg(feature = "resolve-file")]
     async fn test_async_concurrent_retrievals() {
         let mut temp_files = vec![];
         let mut uris = vec![];

--- a/crates/jsonschema/tests/wasm.rs
+++ b/crates/jsonschema/tests/wasm.rs
@@ -1,0 +1,49 @@
+#![cfg(all(target_arch = "wasm32", target_os = "unknown"))]
+
+use jsonschema::{self, Draft};
+use serde_json::json;
+use wasm_bindgen_test::wasm_bindgen_test;
+
+#[wasm_bindgen_test]
+fn validates_simple_object_instances() {
+    let schema = json!({
+        "type": "object",
+        "properties": {
+            "name": { "type": "string" },
+            "age": { "type": "integer", "minimum": 0 }
+        },
+        "required": ["name"]
+    });
+
+    let validator = jsonschema::validator_for(&schema).expect("valid schema");
+
+    assert!(validator.is_valid(&json!({"name": "Ferris", "age": 7})));
+    assert!(!validator.is_valid(&json!({"name": 9, "age": 7})));
+    assert!(!validator.is_valid(&json!({"age": 7})));
+}
+
+#[wasm_bindgen_test]
+fn validates_formats_with_options() {
+    let options = jsonschema::options()
+        .with_draft(Draft::Draft202012)
+        .should_validate_formats(true);
+    let validator = options
+        .build(&json!({"type": "string", "format": "email"}))
+        .expect("schema builds");
+
+    assert!(validator.is_valid(&json!("demo@example.com")));
+    assert!(!validator.is_valid(&json!("definitely not an email")));
+}
+
+#[wasm_bindgen_test]
+fn validates_uuid_format() {
+    let options = jsonschema::options()
+        .with_draft(Draft::Draft202012)
+        .should_validate_formats(true);
+    let validator = options
+        .build(&json!({"type": "string", "format": "uuid"}))
+        .expect("schema builds");
+
+    assert!(validator.is_valid(&json!("67e55044-10b1-426f-9247-bb680e5fe0c8")));
+    assert!(!validator.is_valid(&json!("67e5504410b1426f9247bb680e5fe0c8"))); // missing hyphens
+}

